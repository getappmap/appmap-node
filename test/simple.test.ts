import { spawn } from "node:child_process";
import { cpSync, mkdirSync, readFileSync } from "node:fs";
import { join } from "node:path";

import tmp from "tmp";

import {
  getAppMapBinPath,
  integrationTest,
  readAppmap,
  readAppmaps,
  resolveTarget,
  runAppmapNode,
  runAppmapNodeWithOptions,
  spawnAppmapNode,
  testDir,
} from "./helpers";

const integrationTestSkipOnWindows = process.platform == "win32" ? test.skip : integrationTest;

integrationTest("mapping a simple script", () => {
  expect(runAppmapNode("index.js").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTest("mapping an mjs script", () => {
  expect(runAppmapNode("index.mjs").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTest("mapping js class methods and constructors containing super keyword", () => {
  expect(runAppmapNode("class.js").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTestSkipOnWindows("forwarding signals to the child", async () => {
  const daemon = spawnAppmapNode("daemon.mjs");
  await new Promise<void>((r) =>
    daemon.stdout.on("data", (chunk: Buffer) => chunk.toString().includes("starting") && r()),
  );

  daemon.kill("SIGINT");
  await new Promise((r) => daemon.once("exit", r));

  expect(daemon.exitCode).toBe(42);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTestSkipOnWindows(
  "forwards SIGINT (ctrl-c) properly to the grandchild having active setInterval",
  async () => {
    // This test tests a fix (see: forwardSignals.ts) for a bug which occured
    // when there is a script with active setInterval call and appmap-node
    // runs this script indirectly with a node process: appmap-node node interval.js.
    // We run the commands similarly as reported in:
    // https://github.com/getappmap/appmap-node/issues/118
    const child = spawn("npx", [getAppMapBinPath(), "node", "interval.js"], {
      cwd: resolveTarget(),
      shell: true,
      detached: true,
    });

    await new Promise<void>(
      (r) =>
        child.stdout?.on("data", (chunk: Buffer) => chunk.toString().includes("heartbeat") && r()),
    );

    process.kill(-child.pid!, "SIGINT");

    await new Promise<void>((r) => child.once("exit", r));
    expect(readAppmap()).toMatchSnapshot();
  },
);

integrationTest("mapping generator functions", () => {
  expect(runAppmapNode("generator.js").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTest("mapping a custom Error class with a message property", () => {
  expect(runAppmapNode("inspect.js").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTestSkipOnWindows("finish signal is handled", async () => {
  const server = spawnAppmapNode("server.mjs");
  await new Promise<void>((r) =>
    server.stdout.on("data", (chunk: Buffer) => chunk.toString().includes("starting") && r()),
  );

  // verify that we don't have any half-written AppMaps
  expect(Object.entries(readAppmaps())).toHaveLength(0);

  server.kill("SIGINT");
  await new Promise((r) => server.once("exit", r));

  expect(readAppmap()).toMatchSnapshot();
});

integrationTest("mapping an extensionless CommonJS file", () => {
  const args = ["./extensionless"];
  if (process.platform == "win32") args.unshift("node");
  expect(runAppmapNode(...args).status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

integrationTest("running a script after changing the current directory", () => {
  // Need to make sure the appmap "root" stays the same after
  // appmap-node is run, even if the current directory changes.
  const args =
    process.platform == "win32"
      ? ["cd subproject & node subproject.js"]
      : ["bash", "-c", "cd subproject; node subproject.js"];

  expect(runAppmapNode(...args).status).toBe(0);
  expect(readAppmap()).toBeDefined();
});

integrationTest("mapping a script with tangled async functions", () => {
  expect(runAppmapNode("async.mjs").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
});

const asyncTimeoutCases = new Map<string, string[]>([
  // No async tracking
  ["0", ["1 task", "2 process", "return 2", "return 1", "5 getMessage", "return 5"]],
  // Async tracking with a quick timeout
  ["10", ["1 task", "5 getMessage", "return 5", "2 process", "return 2", "return 1"]],
  // Full async tracking
  ["3000", ["1 task", "2 process", "5 getMessage", "return 5", "return 2", "return 1"]],
]);

for (const [timeout, expectedOrder] of asyncTimeoutCases) {
  integrationTest("mapping a script using async tracking timeout " + timeout, () => {
    const options = {
      env: { ...process.env, APPMAP_ASYNC_TRACKING_TIMEOUT: timeout },
    };
    expect(runAppmapNodeWithOptions(options, "asyncTimeout.mjs").status).toBe(0);
    const appmap = readAppmap();
    expect(
      appmap.events?.map((e) =>
        e.event == "call" && "method_id" in e
          ? `${e.id} ${e.method_id}`
          : `${e.event} ${"parent_id" in e ? e.parent_id : ""}`,
      ),
    ).toEqual(expectedOrder);
    expect(appmap).toMatchSnapshot();
  });
}

integrationTest("creating a default config file", () => {
  const index = resolveTarget("index.js");
  const target = tmp.dirSync({ unsafeCleanup: true });
  mkdirSync(join(target.name, "test"));
  testDir(join(target.name, "test"));

  cpSync(index, resolveTarget("index.js"));

  const cfgPath = resolveTarget("appmap.yml");
  expect(runAppmapNode("index.js").status).toBe(0);
  expect(readAppmap()).toBeDefined();

  // check that the default config file was written
  expect(readFileSync(cfgPath, "utf8")).toMatchSnapshot();
});
