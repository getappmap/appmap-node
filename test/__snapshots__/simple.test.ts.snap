// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping a simple script 1`] = `
{
  "classMap": [
    {
      "children": [
        {
          "children": [
            {
              "location": "./index.js:13",
              "name": "immediatePromise",
              "static": true,
              "type": "function",
            },
            {
              "location": "./index.js:4",
              "name": "foo",
              "static": true,
              "type": "function",
            },
            {
              "location": "./index.js:8",
              "name": "promised",
              "static": true,
              "type": "function",
            },
          ],
          "name": "",
          "type": "class",
        },
      ],
      "name": "index",
      "type": "package",
    },
  ],
  "eventUpdates": {
    "4": {
      "elapsed": 31.337,
      "event": "return",
      "id": 4,
      "parent_id": 3,
      "return_value": {
        "class": "Promise",
        "value": "Promise { 'promised return' }",
      },
      "thread_id": 0,
    },
  },
  "events": [
    {
      "defined_class": "",
      "event": "call",
      "id": 1,
      "lineno": 4,
      "method_id": "foo",
      "parameters": [
        {
          "class": "Number",
          "name": "x",
          "value": "42",
        },
      ],
      "path": "./index.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 2,
      "parent_id": 1,
      "return_value": {
        "class": "Number",
        "value": "84",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "",
      "event": "call",
      "id": 3,
      "lineno": 8,
      "method_id": "promised",
      "parameters": [],
      "path": "./index.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 4,
      "parent_id": 3,
      "return_value": {
        "class": "Promise",
        "value": "Promise { <pending> }",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "",
      "event": "call",
      "id": 5,
      "lineno": 13,
      "method_id": "immediatePromise",
      "parameters": [],
      "path": "./index.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 6,
      "parent_id": 5,
      "return_value": {
        "class": "Promise",
        "value": "Promise { 'immediate' }",
      },
      "thread_id": 0,
    },
  ],
  "metadata": {
    "app": "appmap-node",
    "client": {
      "name": "appmap-node",
      "url": "https://github.com/getappmap/appmap-node",
      "version": "0.0.1",
    },
    "language": {
      "engine": "Node.js",
      "name": "ECMAScript",
      "version": "test node version",
    },
    "name": "test process recording",
    "recorder": {
      "name": "process",
      "type": "process",
    },
  },
  "version": "1.12",
}
`;
